src/server/render-context.js

```diff
@@ -3,11 +3,11 @@
 import { isUndef } from 'shared/util'
 
 type RenderState = {
-  type: 'Element';
+  type: 'Element' | 'Fragment';
   rendered: number;
   total: number;
-  endTag: string;
   children: Array<VNode>;
+  endTag?: string;
```

Comment:
When type is `Element`, `RednerState` definitely has `endTag`. Writing `type` in union type will lose this information.
If we separate it into two state type: one is `Element` and the other is `Fragment`, this information can be kept.